Major Refactor - Request for Comment

GunnerGuyven opened the pull request:
I've completed a major refactor of this project. The main goals were to support additional features that were half-complete or missing before, and make one major design change:

## Design Change

Get off the `emcc` settings treadmill (stop Keeping up with the Joneses)

The original implementation invoked `emcc` directly, but over time the upstream tree-sitter project had altered conventions and these settings had to be updated here to continue working for modern grammars. Instead now, I'm invoking tree-sitter's own official cli to perform compilation which should be much more stable than embedding it's functionality as that project has had a bit of churn in recent years

### Benefits

- a `tree-sitter-cli` node package exists, so we don't even need to have tooling installed locally to use it and upgrades are natural
  - the cli has a 'native' and a 'docker' mode that look for `emcc` to be installed on your system and use that, or use `docker` if you have that to set up an `emcc` instance for you
- additional tree-sitter grammar formats will be supported automatically

## Refined Features

A few of these features were conceptually there, but didn't practically work except in very specific scenarios. They should be ready to go now.

### Building for Production

Some problems that existed were 'wasm' files were served in a different path than they were built to by default, and when building for production, they weren't included (or were overwritten by later build processes). This is possibly a change in `vite` since the plugin's conception.

To correct this, 'wasm' output is put by default not in 'dist' where `vite` is expected to produce it's build, but in '.grammar' (or any other folder of your configuration). A specific step is included during a 'production' build that emits the grammars that are being tracked into the bundle (including 'tree-sitter.wasm' if we're hosting that).

### Display

An issue during normal usage was how long building could take that would result in uncertainty whether `vite` had crashed or it was working normally (and building a grammar). To help with this, I added a message and a spinner that hopefully looks nice and professional.

I also added a bit of logging that can be turned on and off in configuration.

![image](https://github.com/user-attachments/assets/50a61a52-8547-418a-8302-a47625ddfb12)

## Deno, JSR, and Typescript

Perhaps controversially, I have elected to move the build to `deno` (from `node`). Partly this was done because I use `deno` most regularly now instead of `node`, but additionally this was to make publishing easier to the [jsr.io](https://jsr.io/) package repository.

Incidentally, I have done so. The package is published at [vite-plugin-tree-sitter](https://jsr.io/@guyven/vite-plugin-tree-sitter). Please suggest any changes to the README you'd like to see (like an attribution / contribution section where the past devs are mentioned).

A benefit of this is that it is well supported for all of the present major javascript platforms.  I have tested with `deno` and `pnpm` and included those in the '/examples' folder.  I have not yet updated the previous 'solidjs' example but will before long.

## Comments Requested

Please let me know if you would like to receive this update here.  I'm publishing the jsr package out of my fork presently (and probably will continue to do so for my own convenience).  I'm willing to update the README with any information you'd like to add.

I think it is useful to have the updates and releases mirrored here because this project is known and linked around the internet, so it's a way for people to discover this tool who are looking for it.

GunnerGuyven committed:
beginning creation of cli-based replacement

formatting project toward a release on jsr ultimately

GunnerGuyven committed:
porting cli-based plugin

deno used for building

GunnerGuyven committed:
updated readme

added node types for dev
added util.ts to publish include

GunnerGuyven committed:
corrected bad jsr publish

fixed incorrect tool lookup
updated vite to 6
added deno usage example

GunnerGuyven committed:
updated deno example to use jsr

GunnerGuyven committed:
removed ora from example

GunnerGuyven committed:
expanded ability to deal with legacy grammars

attempts to use the tree-sitter tool to upgrade a grammar if unable to
find the expected definition format

GunnerGuyven committed:
added pnpm example

GunnerGuyven committed:
0.2.2

GunnerGuyven committed:
replaced node functionality with deno

GunnerGuyven committed:
updated examples

GunnerGuyven committed:
added deno and node descriminating fs methods

maybe this will simplify compatibility going forward
added trace log level for understanding failure conditions better

GunnerGuyven committed:
0.2.4

GunnerGuyven committed:
corrections for types

GunnerGuyven committed:
more type corrections

GunnerGuyven committed:
correcting whitelist for publish

GunnerGuyven committed:
updated deno example

GunnerGuyven committed:
transitioned to absolute paths in some places

added decode to node shell execution output

GunnerGuyven committed:
0.2.5

GunnerGuyven committed:
inlined call to createReadStream

vite was messing up in node for some reason without doing this

GunnerGuyven committed:
updated examples

GunnerGuyven committed:
updated documentation

GunnerGuyven committed:
updated examples

GunnerGuyven committed:
corrected node production mode output

added readFile to fs layer
removed extra path types from node layer (abandoned feature)
corrected pnpm example

GunnerGuyven committed:
updated examples

milahu commented:
thx 4 your work

sry 4 my delay
[im in jail since 2024-12-09](https://github.com/milahu/alchi/blob/master/deutsch/feedback/knast-2024/2024-12-09.prison-for-book-distributor.md)

> Please let me know if you would like to receive this update here

yes please
youre a contributor already
just help yourself : )

i cant test your work now
but i trust you

GunnerGuyven committed:
added cli interactive mode

milahu commented on fs.ts:
```diff
@@ -0,0 +1,233 @@
+		return Defaults.forceRuntime
+	}
+	const ua = navigator.userAgent
+	const runtime = ua.startsWith("Deno")
+		? Runtime.Deno
+		: ua.startsWith("Node.js")
+			? Runtime.Node
+			: ua.startsWith("Bun")
+				? Runtime.Bun
+				: ua.startsWith("Cloudflare-Workers")
+					? Runtime.CloudflareWorkers
+					: Runtime.unknown
```
dedent

```js
	const runtime =
		ua.startsWith("Deno") ? Runtime.Deno :
		ua.startsWith("Node.js") ? Runtime.Node :
		ua.startsWith("Bun") ? Runtime.Bun :
		ua.startsWith("Cloudflare-Workers") ? Runtime.CloudflareWorkers :
		Runtime.unknown
```

GunnerGuyven replied:
I like this better, and it turns out prettier has an option for this `"experimentalTernaries": true`

milahu commented on fs.ts:
```diff
@@ -0,0 +1,233 @@
+	FSReadFile,
+	FSSpawnShell,
+} from "./types.ts"
+import {
```
this glue code is terrible...
looks like you reinvented https://github.com/cross-org/fs and https://github.com/cross-org/utils

please, keep it simple
the ecosystem fragmentation (js, ts, deno, bun, ...) is bad enough...

GunnerGuyven replied:
I disagree this is terrible, but thanks for pointing out those projects.  At the point that they offer all the features I need (and with native implementations instead of running node-compatibility shims), I'm happy to switch over.  I will probably start moving some functions as I go.

milahu commented on mod.ts:
```diff
@@ -0,0 +1,237 @@
+	const wasmServeList = new Map()
+	wasmServeList.set(
+		"/tree-sitter.wasm",
+		fsPathJoin("node_modules", "web-tree-sitter", "tree-sitter.wasm"),
```
`fsPathJoin(require.resolve("web-tree-sitter"), "tree-sitter.wasm")`

GunnerGuyven replied:
I disliked this as it was, thanks.  If this works in Deno as expected, I'll switch over.

GunnerGuyven replied:
This has been addressed by removing this feature.  Latest web-tree-sitter no longer requires this plugin's help to bundle properly.

GunnerGuyven committed:
added spawnShell for node

updated formatting

GunnerGuyven committed:
updated release documentation

GunnerGuyven committed:
updated deno example

GunnerGuyven committed:
in-code version bump

suppressed warning

GunnerGuyven committed:
removed web-tree-sitter special hosting in plugin

updated dependencies
modified deno-based example
- new web-tree-sitter semantics and config
- removed extra dev deno.json as deno now has the 'links' option

GunnerGuyven committed:
corrected type for pathJoin

GunnerGuyven committed:
updated pnpm example

milahu commented:
looks good, thanks : )
